### PR TITLE
Replace THPLayout with at::Layout in Python Argument Parser (#34543)

### PR DESCRIPTION
**Pull request by terhuhf**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34584 Replace THPLayout with at::Layout in Python Argument Parser (#34543)**

Summary:

Test Plan:
```
python setup.py develop
python test/test_torch.py
```
Output:
```
...
Ran 3834 tests in 198.825s

OK (skipped=180)
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20403330](https://our.internmc.facebook.com/intern/diff/D20403330)

### CODE REVIEW

File: torch/csrc/utils/python_arg_parser.h [hunk 1]
```diff
@@ -162,8 +163,11 @@ struct PythonArgs {
   inline c10::optional<int64_t> toInt64Optional(int i);
   inline c10::optional<bool> toBoolOptional(int i);
   inline c10::optional<double> toDoubleOptional(int i);
-  inline const THPLayout& layout(int i);
-  inline const THPLayout& layoutWithDefault(int i, const THPLayout& default_layout);
+  // inline const at::Layout& layout(int i);
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.h]: nit: we can delete these, right? :)

Comment (terhuhf) [reply torch/csrc/utils/python_arg_parser.h]: Thanks Ed! I also noticed it as well. I'm going to submit the change really quick.

**Comment by terhuhf** [comment]
I'm looking at the test failures. Not sure if they are caused by me :/

**Comment by ezyang** [comment]
> I'm looking at the test failures. Not sure if they are caused by me :/

Yes, they're caused by this diff. I see some suspicious spots in the diff, comments coming.

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -1359,15 +1359,13 @@ def make_python_binding_args(declaration):
         python_binding_arguments.append(dtype_arg)
 
     if is_factory_function or is_like_or_new_function_with_options:
-        py_default_layout = '*torch::getLayout(self.options().backend())' if is_like_or_new_function_with_options else None
         layout_arg = {
-            'default': 'torch.strided',
-            'dynamic_type': 'Layout',
+            'default': 'None',
```
Comment (ezyang) [inline tools/autograd/gen_python_functions.py]: You should preserve the default here; actually, the reason the old PR wasn't landable was because changing the defaults in places like this caused other parts of the codebase to break. It would be pretty cool to fix this (cc @bhosmer) but it's much bigger than a bootcamp task.

More concretely, what you need to do is rewrite `py_default_layout` to produce the C++ version (rather than the Python version) and initialize it using this code. Look at dtype to see how we do this in the other cases.

File: tools/autograd/templates/python_torch_functions.cpp [hunk 3]
```diff
@@ -106,8 +106,8 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
-    "arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "arange(Scalar end, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool pin_memory=False, bool requires_grad=False)",
+    "arange(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool pin_memory=False, bool requires_grad=False)",
```
Comment (ezyang) [inline tools/autograd/templates/python_torch_functions.cpp]: Undo this change

File: tools/autograd/templates/python_torch_functions.cpp [hunk 4]
```diff
@@ -173,7 +173,7 @@ static PyObject * THPVariable_range(PyObject* self, PyObject* args, PyObject* kw
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "range(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
+    "range(Scalar start, Scalar end, Scalar step=1, *, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
```
Comment (ezyang) [inline tools/autograd/templates/python_torch_functions.cpp]: Same here, undo this change

File: tools/autograd/templates/python_torch_functions.cpp [hunk 5]
```diff
@@ -241,8 +240,8 @@ static PyObject * THPVariable_randint(PyObject* self_, PyObject* args, PyObject*
 {
   HANDLE_TH_ERRORS
   static PythonArgParser parser({
-    "randint(int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
-    "randint(int64_t low, int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout layout=torch.strided, Device device=None, bool requires_grad=False)",
+    "randint(int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
+    "randint(int64_t low, int64_t high, IntArrayRef size, *, Generator generator=None, Tensor out=None, ScalarType dtype=None, Layout? layout=None, Device device=None, bool requires_grad=False)",
```
Comment (ezyang) [inline tools/autograd/templates/python_torch_functions.cpp]: And here

File: torch/csrc/utils/python_arg_parser.cpp [hunk 6]
```diff
@@ -413,11 +413,11 @@ void FunctionParameter::set_default_str(const std::string& str) {
     }
   } else if (type_ == ParameterType::LAYOUT) {
     if (str == "None") {
-      default_layout = nullptr;
+      // do nothing
```
Comment (ezyang) [inline torch/csrc/utils/python_arg_parser.cpp]: This is a memory hazard, as when `ParameterType::LAYOUT`, `default_layout` may not be initialized. Maybe add an assert here that it is optional, in this case.

Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.cpp]: (There's actually a lot of hazards here, so https://github.com/pytorch/pytorch/pull/33771 was sort of trying to tighten things up, but I also ran out of time on that PR too :)

File: torch/csrc/utils/python_arg_parser.cpp [hunk 7]
```diff
@@ -413,11 +413,15 @@ void FunctionParameter::set_default_str(const std::string& str) {
     }
   } else if (type_ == ParameterType::LAYOUT) {
     if (str == "None") {
-      default_layout = nullptr;
+      // Assert if default_layout was initialized by checking if the value is
+      // value.
+      TORCH_INTERNAL_ASSERT_DEBUG_ONLY(default_layout == at::Layout::Strided
+          || default_layout == at::Layout::Sparse
+          || default_layout == at::Layout::Mkldnn);
```
Comment (terhuhf) [inline torch/csrc/utils/python_arg_parser.cpp]: @ezyang  Here I just check if the value of the enum is valid. Not sure if this is ok. And not sure if I should assign default_layout a value, like Strided by default, in this case.

File: tools/autograd/gen_python_functions.py [hunk 8]
```diff
@@ -1359,13 +1359,13 @@ def make_python_binding_args(declaration):
         python_binding_arguments.append(dtype_arg)
 
     if is_factory_function or is_like_or_new_function_with_options:
-        py_default_layout = '*torch::getLayout(self.options().backend())' if is_like_or_new_function_with_options else None
+        py_default_layout = '{layout_from_backend(self.options().backend())}' if is_like_or_new_function_with_options else None
```
Comment (terhuhf) [inline tools/autograd/gen_python_functions.py]: oops, added a pair of unnecessary `{}`. Removing them.

File: torch/csrc/utils/python_arg_parser.cpp (hunk 7 above)
Comment (ezyang) [reply torch/csrc/utils/python_arg_parser.cpp]: No, this doesn't make sense. `default_layout` is uninitialized memory at this point in time, you're going to fail ASAN if you do this. I suggest that you should just assert that the argument is "optional" (in which case we will correctly skip looking at `default_layout` if someone tries to grab it) and maybe, for good luck, audit the code gen to make sure that this invariant is respected.

File: tools/autograd/templates/python_torch_functions.cpp [hunk 9]
```diff
@@ -121,13 +121,13 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
       const auto options = TensorOptions()
           .dtype(scalarType)
           .device(r.device(4))
-          .layout(r.layout(3).layout)
+          .layout(r.layoutOptional(3))
```
Comment (ezyang) [inline tools/autograd/templates/python_torch_functions.cpp]: The `Layout` argument here is not optional (there is no question mark after it), so this should be a `r.layout(3)` call. The original PR gives bad guidance here, because it was also trying to change the types of the Layout to be Layout?, but that's no longer in scope for this PR.

Basically, this line should look similar to the dtype/device setters you see above. A bunch of other places need to respect in similar ways.

File: torch/csrc/utils/python_arg_parser.cpp (hunk 6 above)
Comment (terhuhf) [reply torch/csrc/utils/python_arg_parser.cpp]: Fixed with assert allows_none

**Comment by ezyang** [comment]
the test failures are not all your fault; try rebasing to master

File: tools/autograd/gen_python_functions.py [hunk 10]
```diff
@@ -1359,13 +1359,13 @@ def make_python_binding_args(declaration):
         python_binding_arguments.append(dtype_arg)
 
     if is_factory_function or is_like_or_new_function_with_options:
-        py_default_layout = '*torch::getLayout(self.options().backend())' if is_like_or_new_function_with_options else None
+        py_default_layout = 'layout_from_backend(self.options().backend())' if is_like_or_new_function_with_options else None
         layout_arg = {
             'default': 'torch.strided',
-            'dynamic_type': 'Layout',
+            'dynamic_type': 'c10::optional<Layout>',
```
Comment (ezyang) [inline tools/autograd/gen_python_functions.py]: optional or not here?

Comment (terhuhf) [reply tools/autograd/gen_python_functions.py]: @ezyang  quick question, what is the difference between dynamic_type, type, and simple_type?

File: tools/autograd/templates/python_torch_functions.cpp [hunk 11]
```diff
@@ -141,13 +141,13 @@ static PyObject * THPVariable_arange(PyObject* self, PyObject* args, PyObject* k
...
           .device(r.device(6))
-          .layout(r.layout(5).layout)
+          .layout(r.layout(5))
           .requires_grad(r.toBool(8))
           .pinned_memory(r.toBool(7));
       return wrap(dispatch_arange(start, end, step, options));
     } else {
       TORCH_CHECK(!r.toBool(7), " `pin_memory` and `out` parameters are incompatible");
-      check_out_type_matches(r.tensor(3), r.scalartype(4), r.isNone(4), r.layout(5), r.isNone(5),
+      check_out_type_matches(r.tensor(3), r.scalartype(4), r.isNone(4), r.layout(5),
```
Comment (terhuhf) [inline tools/autograd/templates/python_torch_functions.cpp]: @ezyang  Does here look alright? I didn't call `r.isNone` on layout. But `check_out_type_matched` takes an optional of Layout, and in the function it check if it's initialized.

Comment (ezyang) [reply tools/autograd/templates/python_torch_functions.cpp]: Yes, this is fine. The other calls should get updated to also call Optional too, but that's a change for another day.

File: tools/autograd/gen_python_functions.py (hunk 10 above)
Comment (ezyang) [reply tools/autograd/gen_python_functions.py]: I'm not 100% sure. However, I think it's something like: `type` is the C++ type of the argument, complete with references. `simple_type` has all the references stripped off, and `optional` converted into the postfix question mark. And `dynamic_type` says what the actual runtime type is, e.g., `const Tensor&` might be the type, but `Tensor` is the `dynamic_type`. The best way to find out for sure, though, is to read the code responsible for generating these.

**Review by ezyang — APPROVED** [review]
assuming CI passes this LGTM
